src/tracerunner.rs

```diff
@@ -0,0 +1,159 @@
+        // we have to copy over current stack frame from interpreter
+        let mut locals = vec![Value::Null; trace.locals_count];
+        {
+            let interp_locals = &interp.frames.last().unwrap().locals;
+            for idx in 0..interp_locals.len() {
+                locals[idx] = interp_locals[idx].clone();
+            }
+        }
+
+        Runner {
+            interp: interp,
+            trace: &trace.trace,
+            stack: Vec::new(),
+            locals: locals,
+        }
+    }
+
+    pub fn run(&mut self) -> InstrPtr {
+        use TraceInstruction::*;
+
+        let mut pc = 0;
+        loop {
+            let instr = &self.trace[pc];
+            pc = (pc + 1) % self.trace.len();
+
+            info!("TEXEC: {:?}", instr);
+
+            match *instr {
+                Add        => self.add(),
+                Cmp(how)   => self.cmp(how),
+                Load(idx)  => self.load(idx),
+                Store(idx) => self.store(idx),
+                ArrayGet   => self.array_get(),
+                Const(val) => self.stack.push_from(val),
+                Clone      => {}
+
+                Guard(ref guard) => {
+                    if let Err(recovery) = self.check_guard(guard) {
+                        return recovery;
+                    }
+                }
+
+                // these opcodes are not needed for example
+                Array(_) | Push | Print | Len => unimplemented!(),
+            }
+        }
+    }
+
+    fn check_guard(&mut self, guard: &Guard) -> Result<(), InstrPtr> {
+        let check = self.stack.pop_into::<bool>() == guard.condition;
+        check.ok_or_else(||{
+                self.recover(guard);
+                InstrPtr::new(guard.frame.func.clone(), guard.pc)
+            })
+    }
+
+    /// Recovery (aka Blackholing)
+    ///
+    /// Execution has reached a point, where the trace isn't valid anymore.
+    /// The goal is to return to the interpreter, but the state has to be
+    /// recovered first.
+    ///
+    /// The following states have to be recovered:
+    ///     * stack-frames (call-frames)
+    ///       The failed guard might have failed within an inlined function call.
+    ///       Thus, we have to reconstruct all missing callframes, before the
+    ///       the interpreter can gain back control.
+    ///       Second, we also have to consider the frame where the loop resides
+    ///       in, since state might has also has changed there.
+    ///
+    ///     * value stack
+    ///       Also the operand stack has to be recovered.
+    ///       Foremost, the condition, which caused the guard to fail, has to be
+    ///       restored.
+    ///       TODO: Are there other values which might have to be recovered?
+    fn recover(&mut self, guard: &Guard) {
+        // remove the last callframe of the Interpreter
+        // it gets replaced with our updated version
+        self.interp.frames.pop().unwrap();
```

Comment:
I'm a bit surprised that we can discard the current frame entirely. I assumed (maybe wrongly!) that it would have locals (etc.) which would need to be copied in the to-be-created frames?